### PR TITLE
GH2199: Add GlobPattern class

### LINKED ISSUE #2199: Add GlobPattern class
Add GlobPattern class to signal where a glob pattern can be used.

 - Implicitly convertable to string
 - Implicitly convertable from string

### PR DESCRIPTION
**Pull request by devlead**
* fixes #2199

### CODE REVIEW

File: src/Cake.Core/IO/GlobPattern.cs [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+        /// Determines whether two <see cref="GlobPattern"/>. instances are equal.
+        /// </summary>
+        /// <param name="obj">The object to compare with the current object.</param>
+        /// <returns>
+        /// <c>true</c> if the specified object is equal to the current object; otherwise, <c>false</c>.
+        /// </returns>
+        public override bool Equals(object obj)
+        {
+            return obj is GlobPattern pattern &&
+                   Pattern == pattern.Pattern;
```
Comment (patriksvensson) [inline src/Cake.Core/IO/GlobPattern.cs:68]: Shoulf probably do a String.Equals and provide StringComparison.Ordinal here.

File: src/Cake.Core/IO/GlobPattern.cs [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        /// </summary>
+        public string Pattern { get; }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="GlobPattern"/> class.
+        /// </summary>
+        /// <param name="pattern">The globber pattern.</param>
+        public GlobPattern(string pattern)
+        {
+            Pattern = pattern;
```
Comment (patriksvensson) [inline src/Cake.Core/IO/GlobPattern.cs:25]: Should throw if pattern is null here?

**Review by patriksvensson — CHANGES_REQUESTED** [review]

File: src/Cake.Core/IO/GlobPattern.cs [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+        /// <summary>
+        /// Returns the hash code for the <see cref="GlobPattern"/>.
+        /// </summary>
+        /// <returns>
+        /// A 32-bit signed integer hash code.
+        /// </returns>
+        public override int GetHashCode()
+        {
+            return -891421391 + EqualityComparer<string>.Default.GetHashCode(Pattern);
```
Comment (patriksvensson) [inline src/Cake.Core/IO/GlobPattern.cs:79]: No need for the constant here.

**Review by patriksvensson — CHANGES_REQUESTED** [review]

File: src/Cake.Core/IO/GlobPattern.cs (hunk 2 above)
Comment (devlead) [reply src/Cake.Core/IO/GlobPattern.cs:30]: Added null check.

File: src/Cake.Core/IO/GlobPattern.cs (hunk 1 above)
Comment (devlead) [reply src/Cake.Core/IO/GlobPattern.cs]: Switched to `StringComparer.Ordinal.Equals`

File: src/Cake.Core/IO/GlobPattern.cs (hunk 3 above)
Comment (devlead) [reply src/Cake.Core/IO/GlobPattern.cs]: Switched to `StringComparer.Ordinal.GetHashCode`

**Review by patriksvensson — APPROVED** [review]

**Review by gep13 — APPROVED** [review]
LGTM

**Comment by gep13** [comment]
@devlead your changes have been merged, thanks for your contribution 👍

**Pull request merged**